integrators/integrators_body.hpp

```diff
@@ -12,10 +12,86 @@
 
+#define PRINCIPIA_CASE_SLMS(kind, method)                      \
+  case serialization::FixedStepSizeIntegrator::kind:           \
+    return SymmetricLinearMultistepIntegrator<methods::method, \
+                                              typename ODE::Position>()
+
+// We do not deserialize an SPRK per se, but only when it is converted to an
+// SRKN.  The reason is that an SPRK is for a different kind of equation than
+// an SRKN, so the two would return different types.  If we ever need to do
+// this we will need to specialize ReadFromMessage somehow.
+#define PRINCIPIA_CASE_SPRK(kind, method)                                 \
+  case serialization::FixedStepSizeIntegrator::kind:                      \
+    do {                                                                  \
+      CHECK(message.has_composition_method());                            \
+      return SprkAsSrknDeserializer<                                      \
+          ODE,                                                            \
+          methods::method,                                                \
+          methods::method::first_same_as_last>::ReadFromMessage(message); \
+    } while (true)
+
+#define PRINCIPIA_CASE_SRKN(kind, method)                         \
+  case serialization::FixedStepSizeIntegrator::kind:              \
+    return SymplecticRungeKuttaNyströmIntegrator<methods::method, \
+                                                 typename ODE::Position>()
+
 namespace principia {
 namespace integrators {
 namespace internal_integrators {
 
+template<typename ODE, typename Method, bool first_same_as_last>
+class SprkAsSrknDeserializer;
+
+template<typename ODE, typename Method>
+class SprkAsSrknDeserializer<ODE, Method, false> {
+ public:
+  static FixedStepSizeIntegrator<ODE> const& ReadFromMessage(
+      serialization::FixedStepSizeIntegrator const& message);
+};
+
+template<typename ODE, typename Method>
+class SprkAsSrknDeserializer<ODE, Method, true> {
+ public:
+  static FixedStepSizeIntegrator<ODE> const& ReadFromMessage(
+      serialization::FixedStepSizeIntegrator const& message);
+};
+
+template<typename ODE, typename Method>
+FixedStepSizeIntegrator<ODE> const&
+SprkAsSrknDeserializer<ODE, Method, false>::ReadFromMessage(
+    serialization::FixedStepSizeIntegrator const& message) {
+  CHECK_EQ(serialization::FixedStepSizeIntegrator::BA,
+           message.composition_method());
+  return SymplecticRungeKuttaNyströmIntegrator<
+      Method,
+      serialization::FixedStepSizeIntegrator::BA,
+      typename ODE::Position>();
+}
+
+template<typename ODE, typename Method>
+FixedStepSizeIntegrator<ODE> const&
+SprkAsSrknDeserializer<ODE, Method, true>::ReadFromMessage(
+    serialization::FixedStepSizeIntegrator const& message) {
+  switch (message.composition_method()) {
+    case serialization::FixedStepSizeIntegrator::ABA:
+      // TODO(phl): This is using BAB for the ABA case.  Fix once we use C++17.
```

Comment:
Just `LOG(FATAL)` here when trying to deserialize something unsupported, instead of incorrectly deserializing.

Reply:
Done.